File: packages/icp_ws/src/icp/src/icp_serial.cpp
#include <iostream>
#include <vector>
#include <time.h>

#include <Eigen/Dense>

#include "ros/ros.h"
// #include "icp/Lidar.h"
#include <sensor_msgs/PointCloud.h>
#include <geometry_msgs/Point32.h>

sensor_msgs::PointCloud prev_pc;
sensor_msgs::PointCloud curr_pc;

// std::tuple<Eigen::Vector2d, std::vector<Vector2d>> compute_mean(const std::vector<Eigen::Vector2d>& pc) {
Eigen::Vector2d compute_mean(const Eigen::MatrixXd& pc) {
  Eigen::Vector2d center = Eigen::Vector2d::Zero();
  for (int i = 0; i < pc.cols(); ++i) {
    center += pc.col(i);
  }
  center /= pc.cols();
  return center;
}

Eigen::Matrix2d compute_cross_covariance(const Eigen::MatrixXd& P, 
                                        const Eigen::MatrixXd& Q, 
                                        const std::vector<int>& correspondences) {
  Eigen::Matrix2d cov = Eigen::Matrix2d::Zero();
  for (const int& i : correspondences) {
    cov += Q.col(correspondences[i]) * P.col(i).transpose();
  }
  return cov;
}


Eigen::MatrixXd icp_serial(int n_itr) {
    // std::cout << "Success! PC Sizes: " << prev_pc.points.size() << " : " << curr_pc.points.size() << std::endl;    
    Eigen::MatrixXd P;
    Eigen::MatrixXd Q;
    if (curr_pc.points.size() < prev_pc.points.size()) {
        P = Eigen::MatrixXd(2, curr_pc.points.size());
        Q = Eigen::MatrixXd(2, curr_pc.points.size());
    } else {
        P = Eigen::MatrixXd(2, prev_pc.points.size());
        Q = Eigen::MatrixXd(2, prev_pc.points.size());
    }
    /* BEGIN ICP ALGORITHM */  
    bool didFirstItr = false;
    for (int itr = 0; itr < n_itr; itr++) {
        // std::cout << "Made it inside! 1" << std::endl;
        // Find Correspondences
        std::vector<int> correspondences;
        for (int i = 0; i < P.cols(); i++) {
            Eigen::Vector2d p;
            if (!(didFirstItr)) {
                p << curr_pc.points[i].x, curr_pc.points[i].y;
                P.col(i) = p; 
            } else {
                p = P.col(i);
            }
            double min_dist = 1000.0;
            int corr_idx = -1;
            for (int j = 0; j < Q.cols(); j++) {
                Eigen::Vector2d q;
                q << prev_pc.points[j].x, prev_pc.points[j].y;
                if (i == 0) {
                    Q.col(j) = q; 
                }
                double curr_dist = (p - q).norm();
                if (curr_dist < min_dist) {
                    min_dist = curr_dist;
                    corr_idx = j;
                }
            }
            correspondences.push_back(corr_idx);
        }
        // std::cout << "Made it inside! 2" << std::endl;

        // Compute Mean of Point Clouds & Cross-Covariance
        Eigen::Vector2d mu_Q = compute_mean(Q);
        Eigen::Vector2d mu_P = compute_mean(P);
        Eigen::Matrix2d cov = compute_cross_covariance(P,Q,correspondences);

        // SVD Decomposition
        Eigen::JacobiSVD<Eigen::MatrixXd> svd(cov, Eigen::ComputeThinU | Eigen::ComputeThinV);
        Eigen::MatrixXd U = svd.matrixU();
        Eigen::MatrixXd S = svd.singularValues().asDiagonal();
        Eigen::MatrixXd V = svd.matrixV();
        // Eigen::MatrixXd reconstructedMatrix = U * S * V.transpose();

        // Shift P's point cloud
        Eigen::MatrixXd R = U * V.transpose();
        Eigen::VectorXd t = mu_Q - R * mu_P;
        P = R * P + t.replicate(1,P.cols());//t;
        // std::cout << "Made it inside! 8" << std::endl;
        if (!(didFirstItr)) {
            didFirstItr = true;
        }
    }
    /* END ICP ALGORITHM */
    return P;
}

sensor_msgs::PointCloud get_pc_msg(Eigen::MatrixXd P) {
    sensor_msgs::PointCloud pc;
    pc.header.stamp = ros::Time::now();
    pc.header.frame_id = "sonic/lidar_frame";
    for (int i = 0; i < P.cols(); i++) {
        geometry_msgs::Point32 point;
        point.x = P.col(i)[0];
        point.y = P.col(i)[1];
        pc.points.push_back(point);
    }
    return pc;
}

// void lidarMessageCallback(const icp::Lidar::ConstPtr& msg)
void lidarMessageCallback(const sensor_msgs::PointCloud::ConstPtr& pc_msg)
{
    //ROS_INFO("Received custom message: data = %f", msg->distances[0]);
    if (curr_pc.points.empty()) {
        curr_pc = *pc_msg;
    } else {
        prev_pc = curr_pc;
        curr_pc = *pc_msg;
    }
}

int main(int argc, char** argv)
{
    // Initialize the ROS node
    ros::init(argc, argv, "icp_serial");

    // Create a ROS NodeHandle
    ros::NodeHandle nh;
    ros::Rate rate(1); // Hz

    // Subscribe to the custom topic with your custom message type
    ros::Subscriber pc_sub = nh.subscribe<sensor_msgs::PointCloud>("lidar_pointcloud", 10, lidarMessageCallback);

    // Publisher of Alligned Point Cloud
    ros::Publisher pc_pub = nh.advertise<sensor_msgs::PointCloud>("icp_pointcloud", 1);

    // ROS spin to wait for messages
    struct timespec start_time, end_time;
    Eigen::MatrixXd P;
    sensor_msgs::PointCloud icp_msg;
    int icp_itr;
    while (ros::ok()) {
        if (!(prev_pc.points.empty()) && !(curr_pc.points.empty())) {
            if (!nh.getParam("icp_itr", icp_itr))
            {
                ROS_ERROR("Failed to get parameter 'icp_itr'. Using default value of 1.");
                icp_itr = 1;  // Default value
            }
            clock_gettime(CLOCK_MONOTONIC, &start_time);
            P = icp_serial(icp_itr);
            clock_gettime(CLOCK_MONOTONIC, &end_time);
            long double elapsed_time = (end_time.tv_sec - start_time.tv_sec) + (end_time.tv_nsec - start_time.tv_nsec) / 1e9L;
            std::cout << "Execution time: " << elapsed_time << " seconds" << std::endl;
            icp_msg = get_pc_msg(P);
            pc_pub.publish(icp_msg);
        }
        ros::spinOnce();
        rate.sleep();
    }
    //   ros::spin();

    return 0;
}

File: packages/icp_ws/src/icp/src/icp_serial.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud
from geometry_msgs.msg import Point32
from sensor_msgs.msg import ChannelFloat32
import numpy as np
from geometry_msgs.msg import Transform


class LidarPointCloudSubscriber:
    def __init__(self):
        # Initialize the ROS node
        rospy.init_node('icp_gauss_newton_py', anonymous=True)

        # Create a subscriber for the PointCloud topic
        self.pointcloud_sub = rospy.Subscriber('lidar_pointcloud', PointCloud, self.pointcloud_callback)
        self.pointcloud_pub = rospy.Publisher('lidar_aligned', PointCloud, queue_size=1)
        self.transform_pub = rospy.Publisher('transform', Transform, queue_size=10)
        self.prev_points = None

    # P is moved data, q is prev data
    def icp_svd(self, p, q):
        p = np.array(p)
        q = np.array(q)

        # Convert p and q to NumPy arrays without the z coordinate
        p = np.array([[point.x, point.y] for point in p], dtype=float).T
        q = np.array([[point.x, point.y] for point in q], dtype=float).T

        # Reshape to ensure that the input arrays have the shape (dimension, num_points)
        if p.ndim == 1:
            p = p.reshape(-1, 1)
        if q.ndim == 1:
            q = q.reshape(-1, 1)

        # Don't love this implementation, but it'll have to do for now
        # chop off the longer array to match the shorter one
        if p.shape[1] > q.shape[1]:
            p = p[:, :q.shape[1]]
        elif q.shape[1] > p.shape[1]:
            q = q[:, :p.shape[1]]

        icp_itr_default = 5
        # Get the parameter 'icp_itr' from the ROS parameter server
        icp_itr = rospy.get_param("icp_itr", icp_itr_default)

        # Check if the parameter retrieval was successful
        if icp_itr is None:
            rospy.logerror("Failed to get parameter 'icp_itr'. Using default value of %d.", icp_itr_default)
            icp_itr = icp_itr_default
        kernel=lambda diff: 1.0
        
        def get_correspondence_indices(P, Q):
            """For each point in P find closest one in Q."""
            p_size = P.shape[1]
            q_size = Q.shape[1]
            correspondences = []
            for i in range(p_size):
                p_point = P[:, i]
                min_dist = 9999
                chosen_idx = -1
                for j in range(q_size):
                    q_point = Q[:, j]
                    dist = np.linalg.norm(q_point - p_point)
                    if dist < min_dist:
                        min_dist = dist
                        chosen_idx = j
                correspondences.append((i, chosen_idx, min_dist))
            return correspondences

        def compute_cross_covariance(P, Q, correspondences, kernel=lambda diff: 1.0):
            cov = np.zeros((2, 2))
            exclude_indices = []
            for i, j, k in correspondences:
                p_point = P[:, [i]]
                q_point = Q[:, [j]]
                weight = kernel(p_point - q_point)
                if weight < 0.01: exclude_indices.append(i)
                cov += weight * q_point.dot(p_point.T)
            return cov, exclude_indices
        
        def center_data(data, exclude_indices=[]):
            reduced_data = np.delete(data, exclude_indices, axis=1)
            center = np.array([reduced_data.mean(axis=1)]).T
            return center

        """Perform ICP using SVD."""
        center_of_Q = center_data(q)
        norm_values = []
        P_values = [p.copy()]
        P_copy = p.copy()
        corresp_values = []
        exclude_indices = []
        R_tot = np.eye(2)
        t_tot = np.zeros((2, 1))
        for i in range(icp_itr):
            center_of_P = center_data(P_copy, exclude_indices=exclude_indices)
            correspondences = get_correspondence_indices(p, q)
            corresp_values.append(correspondences)
            norm_values.append(np.linalg.norm(p - q))
            cov, exclude_indices = compute_cross_covariance(p, q, correspondences, kernel)
            U, S, V_T = np.linalg.svd(cov)
            R = U.dot(V_T)  
            t = center_of_Q - R.dot(center_of_P)
            R_tot = R @ R_tot
            t_tot = R.dot(t_tot + t) # Could be one of these two methods
            # t_tot = R.dot(t_tot) + t
            P_copy = R.dot(P_copy) + t
            P_values.append(P_copy)
        corresp_values.append(corresp_values[-1])
        return R_tot, t_tot, P_copy

    def icp_gauss_newton(self, points, prev_points):
        pass

    def publish_transform(self, rotation_matrix, translation_vector):
        # Create a Transform message
        transform_msg = Transform()
        yaw = np.arctan2(rotation_matrix[1, 0], rotation_matrix[0, 0])
        roll = 0
        pitch = 0
        qx = 0
        qy = 0
        qz = np.sin(yaw/2)
        qw = np.cos(yaw/2)
 
        transform_msg.rotation.x = qx
        transform_msg.rotation.y = qy
        transform_msg.rotation.z = qz
        transform_msg.rotation.w = qw

        # Set the translation vector (3x1 part of the transformation)
        transform_msg.translation.x = translation_vector[0][0]
        transform_msg.translation.y = translation_vector[1][0]
        transform_msg.translation.z = 0

        # Publish the Transform message
        self.transform_pub.publish(transform_msg)


    def pointcloud_callback(self, pointcloud_msg):
        # Extract information from the received PointCloud message
        header = pointcloud_msg.header
        points = pointcloud_msg.points
        channels = pointcloud_msg.channels

        # Process and print the received data (example: print the first 10 points)
        # print(f"Received PointCloud with {len(points)} points:")
        # for i in range(min(10, len(points))):
        #     point = points[i]
        #     intensity = channels[0].values[i] if channels else None
        #     print(f"Point {i + 1}: ({point.x}, {point.y}, {point.z}), Intensity: {intensity}")
        
        if self.prev_points is not None:
            start_time = rospy.Time.now()
            R, t, aligned_points = self.icp_svd(points, self.prev_points)
            end_time = rospy.Time.now()
            time_taken = end_time - start_time
            # convert to seconds
            time_taken = time_taken.secs + time_taken.nsecs * 1e-9
            print(f"Rotation:{R}\nTranslation:{t}\nTime taken:{time_taken}")
            # create pointcloud message
            pointcloud_msg = PointCloud()
            pointcloud_msg.header.stamp = rospy.Time.now()
            pointcloud_msg.header.frame_id = "sonic/lidar_frame"
            print("Aligned Points shape: ", aligned_points.shape)
            points = [Point32(x=float(aligned_points[0][i]), y=float(aligned_points[1][i]), z=0.0) for i in range(aligned_points.shape[1])]
            pointcloud_msg.points = points

            # Add intensity values as an additional channel with default value 1
            intensity_channel = ChannelFloat32()
            intensity_channel.name = "intensity"
            intensity_channel.values = [1.0] * len(points)
            pointcloud_msg.channels.append(intensity_channel)

            self.pointcloud_pub.publish(pointcloud_msg)
            self.publish_transform(R, t)
        self.prev_points = points

if __name__ == '__main__':
    try:
        lidar_subscriber = LidarPointCloudSubscriber()
        rospy.spin()  # Keep the program from exiting
    except rospy.ROSInterruptException:
        pass
